JS Learning Area: update guidance on const and let (#10421)
mdn/content: files/en-us/learn/javascript/first_steps/variables/index.md

This is probably what you'll do most of the time, as it is quicker than doing the two actions on two separate lines.

## A note about var

You'll probably also see a different way to declare variables, using the `var` keyword:

```js
var myName;
var myAge;
```

> Back when JavaScript was first created, this was the only way to declare variables. The design of `var` has issues that make it easy for errors to creep into your code. So `let` was created in modern versions of JavaScript, a new keyword for creating variables that works somewhat differently to `var`, fixing its issues in the process.

Reviewer: ```suggestion
Back when JavaScript was first created, this was the only way to declare variables. The design of `var` is confusing and error-prone. So `let` was created in modern versions of JavaScript, a new keyword for creating variables that works somewhat differently to `var`, fixing its issues in the process.
```

Reviewer: the errors creeping into code kind of sounded like spiders rather than human error ;)

Again, this is a sensible language decision. There is no reason to redeclare variables — it just makes things more confusing.

> For these reasons and more, we recommend that you use `let` in your code, rather than `var`. There is no reason to use `var`, unless you need to support old versions of Internet Explorer with your code (it doesn't support `let` until version 11; the modern Microsoft Edge browser supports `let` just fine).

Reviewer: ```suggestion
For these reasons and more, we recommend that you use `let` in your code, rather than `var`. There is no reason to use `var`, unless you need to support Internet Explorer 10 or older with your code.
```

Reviewer: not sure if we want to start weaning ourselves off of always mentioning Edge when we mention IE, but I am all for it.

## Constants in JavaScript

As well as variables, you can declare constants. These are like variables, except that:

* you must initialize them when you declare them
* you can't assign them a new value after you've initialized them.

For example, using `let` you can declare a variable without initializing it:

```js
let count;
```

If you try to do this using `const` you will see an error:

```js example-bad
const count;
```

> Similarly, with `let` you can initialize a variable, and then assign it a new value:

Author: ```suggestion
+Similarly, with `let` you can initialize a variable, and then assign it a new value (this is also called *reassigning* the variable):
```

const count = 1;
count = 2;
```

Note that although a constant in JavaScript must always name the same value, you can change the content of the value that it names. This isn't a useful distinction for simple types like numbers or booleans, but consider an object:

```js
const bird = { species : 'Kestrel'};
console.log(bird.species);  // "Kestrel"
```

You can update, add, or remove properties of an object declared using `const`, because even though the content of the object has changed, the constant is still pointing to the same object:

```js
bird.species = 'Striated Caracara';
console.log(bird.species);  // "Striated Caracara"
```

## When to use const and when to use let

If you can't do as much with `const` as you can with `let`, why would you prefer to use it rather than `let`? In fact `const` is very useful. If you use `const` to name a value, it tells anyone looking at your code that this name will never be assigned to a different value. Any time they see this name, they will know what it refers to.

In this course, we adopt the following principle about when to use `let` and when to use `const`:

*Use `const` when you can, and use `let` when you have to.*

> This means that if you can initialize a variable when you declare it, and don't need to reassign it later, make it a constant.

Reviewer: ```suggestion
This means that if you can initialize a variable when you declare it and don't need to change it later, make it a constant.
```

Reviewer: not sure if non-native speakers understand reassign as well as change

Author: As said in the call, I'd really like to keep "reassign" here because "change" is ambiguous. I appreciate that "reassign" is not very friendly though. Maybe I could add a bit further up that introduces the word.
